src/kernel/mempool_removal_reason.h

```diff
@@ -5,20 +5,61 @@
 #ifndef BITCOIN_KERNEL_MEMPOOL_REMOVAL_REASON_H
 #define BITCOIN_KERNEL_MEMPOOL_REMOVAL_REASON_H
 
+#include <primitives/transaction.h>
+#include <uint256.h>
+
 #include <string>
+#include <variant>
 
-/** Reason why a transaction was removed from the mempool,
+/** Reasons why a transaction was removed from the mempool,
  * this is passed to the notification signal.
  */
-enum class MemPoolRemovalReason {
-    EXPIRY,      //!< Expired from mempool
-    SIZELIMIT,   //!< Removed in size limiting
-    REORG,       //!< Removed for reorganization
-    BLOCK,       //!< Removed for block
-    CONFLICT,    //!< Removed for conflict with in-block transaction
-    REPLACED,    //!< Removed for replacement
+
+struct ExpiryReason {
+    std::string toString() const noexcept {
+        return "expiry";
+    }
+};
+
+struct SizeLimitReason {
+    std::string toString() const noexcept {
+        return "sizelimit";
+    }
+};
+
+struct ReorgReason {
+    std::string toString() const noexcept {
+        return "reorg";
+    }
+};
+
+struct BlockReason {
+    std::string toString() const noexcept {
+        return "block";
+    }
 };
 
+struct ConflictReason {
+    uint256 conflicting_block_hash;
+    unsigned int conflicting_block_height;
```

Comment:
in https://github.com/bitcoin/bitcoin/pull/29680/commits/7debac0b09faf051229a4ead0b57e6658e00ac11 ("Change MemPoolRemovalReason to variant type"):

I haven't finished reviewing the later commits yet, but seems odd to pass both `conflicting_block_hash` and `conflicting_block_height`. Seems like we should be able to only use `conflicting_block_hash`?

Reply:
The `TxStateBlockConflicted` requires both that's why I added the both of them